src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs

```diff
@@ -1166,6 +1167,239 @@ public bool TryExportCertificatePem(Span<char> destination, out int charsWritten
+        ///       </item>
+        ///     </list>
+        ///   </para>
+        ///   <para>
+        ///     This method does not convert non-ASCII hostnames to the IDNA representation. For Unicode domains,
+        ///     the caller must make use of <see cref="System.Globalization.IdnMapping"/> or an equivalent IDNA mapper.
+        ///   </para>
+        ///   <para>
+        ///     The "exact" matches performed by this routine are <see cref="StringComparison.OrdinalIgnoreCase"/>,
+        ///     as domain names are not case-sensitive.
+        ///   </para>
+        ///   <para>
+        ///     This method does not determine if the hostname is authorized by a trusted authority.  A trust
+        ///     decision cannot be made without additionally checking for trust via <see cref="X509Chain"/>.
+        ///   </para>
+        ///   <para>
+        ///     This method does not check that the certificate has an <c>id-kp-serverAuth</c> (1.3.6.1.5.5.7.3.1)
+        ///     extended key usage.
+        ///   </para>
+        /// </remarks>
+        /// <exception cref="CryptographicException">
+        ///   <para>The certificate contains multiple Subject Alternative Name extensions.</para>
+        ///   <para>- or -</para>
+        ///   <para>The Subject Alternative Name extension or Subject Name could not be decooded.</para>
+        /// </exception>
+        public bool MatchesHostname(string hostname, bool allowWildcards = true, bool allowCommonName = true)
+        {
+            ArgumentNullException.ThrowIfNull(hostname);
+
+            if (hostname.Length == 0)
+            {
+                return false;
+            }
+
+            X509Extension? rawSAN = null;
+
+            foreach (X509Extension extension in Pal.Extensions)
+            {
+                if (extension.Oid!.Value == Oids.SubjectAltName)
+                {
+                    if (rawSAN is null)
+                    {
+                        rawSAN = extension;
+                    }
+                    else
+                    {
+                        throw new CryptographicException(SR.Cryptography_X509_TooManySANs);
+                    }
+                }
+            }
+
+            if (rawSAN is not null)
+            {
+                var san = new X509SubjectAlternativeNameExtension();
+                san.CopyFrom(rawSAN);
+
+                bool hadAny = false;
+
+                if (IPAddress.TryParse(hostname, out IPAddress? ipAddress))
+                {
+                    foreach (IPAddress sanEntry in san.EnumerateIPAddresses())
+                    {
+                        if (sanEntry.Equals(ipAddress))
+                        {
+                            return true;
+                        }
+
+                        hadAny = true;
+                    }
+                }
+                else
+                {
+                    ReadOnlySpan<char> match = hostname;
```

Comment:
I think something missing here is "Is this a valid hostname" check. Right now it takes arbitrary strings that are not RFC 1034 (or whatever RFC you thing is worth following). This could lead to injection / or confusion in `MatchesHostName` as I described in my comment for the tests.

Reply:
I would expect an argument exception if the hostname is not a well-formed hostname.